### ISSUE TITLE
ty doesn't detect same name defined multiple types (aka mypy no-redef)

### ISSUE THREAD

**Issue by karlicoss** [astral-sh/ty#835]
## What happens
If a function is declared twice, ty happily accepts it and infers the type as of the latest definition:

```
def x() -> int:
    return 1

def x() -> str:
    return ""

reveal_type(x)
# Revealed type: `def x() -> str` (revealed-type) [Ln 7, Col 13]
```

ty playground link: https://play.ty.dev/9143957d-9d8d-4f20-b8c0-f6ce700d69f8

## Expected behaviour

Should probably complain (especially if the types are mismatching).

For the same code, mypy complains with [`no-redef` error](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-each-name-is-defined-once-no-redef)

 https://mypy-play.net/?mypy=latest&python=3.12&gist=592caaea970d4463db57c27616547448

Apologies if this is already discussed somewhere, but couldn't find!

### Version

latest git (playground)

**Comment by carljm** [comment]
I think we see this as more of a lint rule, not a core type checker feature. This code doesn't have any type errors in it.

I do think a "this function/class was defined and then shadowed without ever being used" lint rule would make sense to add on top of ty in the future.

**Comment by karlicoss** [comment]
Ah thanks! Out of interest, is there a special lint mode planned for ty? Or you mean it would be part of ruff?

**Comment by carljm** [comment]
The details are still somewhat TBD, but in the long term we envision unifying ruff and ty in some way such that we can have lint rules powered by ty's type analysis.
